### ISSUE TITLE
Per-batch device synchronization stalls scoring and inference

### ISSUE THREAD

**Issue by mmmfrieddough** [mmmfrieddough/anthro-chess#166]
## Goal

Stop the scoring and inference paths from forcing a blocking device
synchronization on every batch. This is waste rather than a tradeoff: the checks
involved can stay exactly as strict.

## What was observed

`anthro eval run` was pointed at `training-blitz-30k-v4` step 8000 over the
canonical view of the 50,073-game frozen pool on Apple Silicon MPS. It ran for
**4 h 05 m** without completing and was killed. Its log shows it had finished the
base and shuffled-rating conditioning arms and started the constant-rating arm,
so the benchmark makes three passes over the pool — 150,219 game scorings — and
four hours covered roughly two of them.

Sampling the live process for five seconds put **693 of 1,300 main-thread
samples** in a single path:

```
slot_nb_bool -> THPVariable_is_nonzero -> item() -> _local_scalar_dense_mps
  -> mps_copy_ -> copy_and_sync -> MPSStream::synchronize
    -> [_MTLCommandBuffer waitUntilCompleted] -> __psynch_cvwait
```

That is Python evaluating a tensor in boolean context. Each occurrence drains
the entire Metal command buffer before Python continues, so no work overlaps.

## Where it comes from

`src/anthro_chess/evaluation/policy.py` — `_validate_logits` ends with
`if not torch.all(torch.isfinite(logits[batch.action_loss_mask]))`. `torch.all`
returns a zero-dimensional tensor and `not` invokes `__bool__` on it. It is
called from `_active_batch` on every batch of the scoring loop.

The same shape is in the per-move runtime path in
`src/anthro_chess/inference/runner.py`, where the finite check on action logits
runs for every generated move. That path is shared by the rollout, ladder, and
termination benchmarks.

In both places the surrounding code transfers the same tensor to the CPU
immediately afterward — the masked active logits in the first case, the detached
logits in the second. So the boolean check pays for a device synchronization
next to a transfer that was already going to happen, and in the scoring path the
masked gather `logits[batch.action_loss_mask]` is computed twice per batch, once
for the check and once for the value that is kept.

## Scope

- Remove the per-batch and per-move synchronizations from the scoring and
  inference hot paths without weakening what the checks reject.
- Remove the duplicated masked gather in the scoring path.
- Confirm the checks still fail on non-finite logits, with coverage that would
  have caught a silent removal.
- Re-measure `anthro eval run` on the same checkpoint and pool view, and the
  efficiency benchmark's throughput-by-batch-size curve, reporting before and
  after.

## Notes

Deliberately separate from #164. That issue buys speed by giving up measurement
strength and is sequenced behind the CUDA move; this gives up nothing and is
fixable and measurable on Apple Silicon today. Its result may change how much of
#164 is left to do, and it should land first.

There is a hypothesis worth testing rather than assuming: the efficiency
benchmark reports 58, 85, 93, and 98 decisions/s at batch sizes 1, 4, 8, and 16.
A blocking synchronization per forward pass would produce diminishing returns of
roughly that shape, because a larger batch can only amortize the stall rather
than overlap around it. If that is the cause, the curve should straighten once
the synchronization is gone. If it does not, the flat curve is a separate
finding and belongs in its own issue.

The three-pass structure of the held-out benchmark is not in scope here. Whether
scoring the pool once per conditioning arm is the right shape is a question for
#164.

## Acceptance criteria

- Neither the scoring loop nor the per-move inference path evaluates a device
  tensor in boolean context.
- Non-finite action logits are still rejected, with tests that fail if the check
  is removed.
- The scoring path computes its masked gather once per batch.
- Before-and-after timings are reported for `anthro eval run` over the canonical
  pool view on one checkpoint, and for the throughput-by-batch-size curve.

## References

- `docs/evaluation.md`
- `docs/training-and-runtime.md`
- #164 — sweep cost reduction, which this precedes
- #162 — the suite driver, which will run these paths repeatedly

**Comment by mmmfrieddough** [comment]
Claimed for implementation by Claude Code.
